### PR TITLE
HUDs can now only change status with sufficient ID access

### PR DESCRIPTION
**Pull request by Danchi299**
## What Does This PR Do
Changing security/medical status of a person now requires ID access of the same consoles.
Medical HUD can now see and change Mental status of a person just like the console counterpart.
Other small HUD adjustments that can be seen in Changelog.

## Why It's Good For The Game
If random assistant gets a Security HUD they won't be able to set everyone they see to arrest due to lack of ID access but they still will be able to read the criminal records of crew. 
Which means RnD is now free to distribute all any any kinds of HUDs without causing paperwork havoc. 
Of course that also means that antagonists won't be able to remove wanted status from themself that easily, 
But if you got a security HUD from officer, shouldn't you already have ID with sufficient access?
And to counteract this Chameleon SecHUD has write access by default without needing ID access.
And yes, Blueshield and IAA still can't change security status of people, now without needing special HUD subtype for it.

Changing and seeing mental status of people is something I did not know existed until I checked medical console,
Making it adjustable in MedHUD should make Psychiatrist job a bit easier and more interactive with rest of medbay.

## Testing
Joined as Blueshield, tried changing my Mental status which succeded, tried changing my Security status which failed.
Joined as AI, checked that all HUDs worked properly.
Ghosted, checked medical HUDs on examine.

## Changelog
:cl:
add: Medical HUDs can now see and change mental status of people.
tweak: Changing records using HUDs can now only be done with sufficient ID access.
tweak: Chameleon HUD now can by default change security status of other people.
tweak: Ghosts can now read medical records of people.
del: Removed read only sunglasses Blueshield and IAA used.
/:cl:

### CODE REVIEW

File: code/__DEFINES/hud.dm [hunk 1]
```diff
@@ -65,5 +65,6 @@
 // on-screen icons, but rather go to examine text.
 #define EXAMINE_HUD_SECURITY_READ "security_read"
 #define EXAMINE_HUD_SECURITY_WRITE "security_write"
-#define EXAMINE_HUD_MEDICAL "medical"
-#define EXAMINE_HUD_SKILLS "skills"
+#define EXAMINE_HUD_MEDICAL_READ "medical_read"
+#define EXAMINE_HUD_MEDICAL_WRITE "medical_write"
+#define EXAMINE_HUD_SKILLS "skills"
```
Comment (Danchi299) [inline code/__DEFINES/hud.dm:70]: Trailing new line?
```suggestion
#define EXAMINE_HUD_SKILLS "skills"

```

**Comment by matttheficus** [comment]
Could you ensure that the Cham SecHUD can still change statuses regardless of access?

The item becomes a bit useless otherwise.

**Comment by Danchi299** [comment]
> Could you ensure that the Cham SecHUD can still change statuses regardless of access?
> 
> The item becomes a bit useless otherwise.

Is already insured.
```dm
/obj/item/clothing/glasses/hud/security/chameleon
	examine_extensions = list(EXAMINE_HUD_SECURITY_READ, EXAMINE_HUD_SECURITY_WRITE)
	flash_protect = FLASH_PROTECTION_FLASH
```
> tweak: Chameleon HUD now can by default change security status of other people.

does not apply to SecHUD implant, still need opinion on that

**Comment by matttheficus** [comment]
I’m fine with the implant by-default having Write access, it takes a bit more effort to get an implant than it does a HUD eyepiece.

File: code/modules/mob/living/carbon/human/human_examine.dm [hunk 2]
```diff
@@ -219,19 +195,48 @@
...
 					if(R.fields["id"] == E.fields["id"])
 						medical = R.fields["p_stat"]
+						mental = R.fields["m_stat"]
 
-		msg += "<span class = 'deptradio'>Physical status:</span> <a href='?src=[UID()];medical=1'>\[[medical]\]</a>\n"
+		var/medical_status = hasHUD(user, EXAMINE_HUD_MEDICAL_WRITE) ? "<a href='?src=[UID()];medical=1'>\[[medical]\]</a>" : "\[[medical]\]"
+		var/mental_status = hasHUD(user, EXAMINE_HUD_MEDICAL_WRITE) ? "<a href='?src=[UID()];mental=1'>\[[mental]\]</a>" : "\[[mental]\]"
+		msg += "<span class = 'deptradio'>Physical status: </span>[medical_status]\n"
+		msg += "<span class = 'deptradio'>Mental Status: </span>[mental_status]\n"
 		msg += "<span class = 'deptradio'>Medical records:</span> <a href='?src=[UID()];medrecord=`'>\[View\]</a> <a href='?src=[UID()];medrecordadd=`'>\[Add comment\]</a>\n"
```
Comment (Burzah) [inline code/modules/mob/living/carbon/human/human_examine.dm:214]:
```suggestion
		msg += "<span class='deptradio'>Physical status: </span>[medical_status]\n"
		msg += "<span class='deptradio'>Mental Status: </span>[mental_status]\n"
		msg += "<span class='deptradio'>Medical records:</span> <a href='?src=[UID()];medrecord=`'>\[View\]</a> <a href='?src=[UID()];medrecordadd=`'>\[Add comment\]</a>\n"
```
I know this is part of the copy paste, but can you fix this class attribute spacing while we are here?

File: code/modules/mob/living/carbon/human/human_examine.dm [hunk 3]
```diff
@@ -219,19 +195,48 @@
...
+								commentLatest = LAZYACCESS(comments, comments.len) //get the latest entry from the comment log
+								if(islist(commentLatest))
+									commentLatest = "[commentLatest["header"]]: [commentLatest["text"]]"
+							else
+								commentLatest = "No entries." //If present but without entries (=target is recognized crew)
+
+			var/criminal_status = hasHUD(user, EXAMINE_HUD_SECURITY_WRITE) ? "<a href='?src=[UID()];criminal=1'>\[[criminal]\]</a>" : "\[[criminal]\]"
+			msg += "<span class = 'deptradio'>Criminal status:</span> [criminal_status]\n"
+			msg += "<span class = 'deptradio'>Security records:</span> <a href='?src=[UID()];secrecordComment=`'>\[View comment log\]</a> <a href='?src=[UID()];secrecordadd=`'>\[Add comment\]</a>\n"
+			msg += "<span class = 'deptradio'>Latest entry:</span> [commentLatest]\n"
```
Comment (Burzah) [inline code/modules/mob/living/carbon/human/human_examine.dm:238]:
```suggestion
			msg += "<span class='deptradio'>Criminal status:</span> [criminal_status]\n"
			msg += "<span class='deptradio'>Security records:</span> <a href='?src=[UID()];secrecordComment=`'>\[View comment log\]</a> <a href='?src=[UID()];secrecordadd=`'>\[Add comment\]</a>\n"
			msg += "<span class='deptradio'>Latest entry:</span> [commentLatest]\n"
```
Same as above.

File: code/modules/mob/living/carbon/examine.dm [hunk 4]
```diff
@@ -314,15 +314,23 @@
 		if(CIH?.examine_extensions)
 			have_hudtypes += CIH.examine_extensions
 
+		var/user_accesses = M.get_access()
+		var/secwrite = has_access(null, list(ACCESS_SECURITY, ACCESS_FORENSICS_LOCKERS), user_accesses) // same as obj/machinery/computer/secure_data/req_one_access
+		var/medwrite = has_access(null, list(ACCESS_MEDICAL, ACCESS_FORENSICS_LOCKERS), user_accesses) // same access as obj/machinery/computer/med_data/req_one_access
+		if(secwrite)
+			have_hudtypes += EXAMINE_HUD_SECURITY_WRITE
+		if(medwrite)
+			have_hudtypes += EXAMINE_HUD_MEDICAL_WRITE
```
[the remark this answers is not in the document]
Comment (Danchi299) [reply code/modules/mob/living/carbon/examine.dm:323]: we do not
but putting a manual list inside a function inside a conditional doesn't seem like a good code styling idea either

[the remark this answers is not in the document]
Comment (DGamerL) [reply code/modules/mob/living/carbon/examine.dm:323]: Fair enough, LGTM

**Review by DGamerL — APPROVED** [review]

**Review by Burzah — APPROVED** [review]

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 5]
```diff
@@ -795,8 +795,31 @@
...
+					for(var/datum/data/record/R in GLOB.data_core.general)
+						if(R.fields["id"] == E.fields["id"])
+							var/setmental = input(usr, "Specify a new mental status for this person.", "Medical HUD", R.fields["m_stat"]) in list("*Insane*", "*Unstable*", "*Watch*", "Stable", "Cancel")
+
+							if(hasHUD(usr, EXAMINE_HUD_MEDICAL_WRITE))
+								if(setmental != "Cancel")
+									R.fields["m_stat"] = setmental
+									modified = 1
+									spawn()
+										sec_hud_set_security_status()
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/human_mob.dm:816]: Is this supposed to reference the security status here?

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 6]
```diff
@@ -795,8 +795,31 @@
 			if(!modified)
 				to_chat(usr, "<span class='warning'>Unable to locate a data core entry for this person.</span>")
 
+	if(href_list["mental"])
+		if(hasHUD(usr, EXAMINE_HUD_MEDICAL_WRITE))
+			if(usr.incapacitated())
+				return
+			var/modified = 0
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/human_mob.dm:802]:
```suggestion
			var/modified = FALSE
```

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 7]
```diff
@@ -795,8 +795,31 @@
...
+			for(var/datum/data/record/E in GLOB.data_core.general)
+				if(E.fields["name"] == perpname)
+					for(var/datum/data/record/R in GLOB.data_core.general)
+						if(R.fields["id"] == E.fields["id"])
+							var/setmental = input(usr, "Specify a new mental status for this person.", "Medical HUD", R.fields["m_stat"]) in list("*Insane*", "*Unstable*", "*Watch*", "Stable", "Cancel")
+
+							if(hasHUD(usr, EXAMINE_HUD_MEDICAL_WRITE))
+								if(setmental != "Cancel")
+									R.fields["m_stat"] = setmental
+									modified = 1
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/human_mob.dm:814]:
```suggestion
									modified = TRUE
```

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 8]
```diff
@@ -782,7 +782,7 @@
 						if(R.fields["id"] == E.fields["id"])
 							var/setmedical = input(usr, "Specify a new medical status for this person.", "Medical HUD", R.fields["p_stat"]) in list("*SSD*", "*Deceased*", "Physically Unfit", "Active", "Disabled", "Cancel")
 
-							if(hasHUD(usr, EXAMINE_HUD_MEDICAL))
+							if(hasHUD(usr, EXAMINE_HUD_MEDICAL_WRITE))
 								if(setmedical != "Cancel")
 									R.fields["p_stat"] = setmedical
 									modified = 1
```
Comment (Burzah) [inline code/modules/mob/living/carbon/human/human_mob.dm:788]:
```suggestion
									modified = TRUE
```
Might as well change these ones as well, while we are here.

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 9]
```diff
@@ -770,7 +770,7 @@
 		add_comment(usr, "security", sanitized)
 
 	if(href_list["medical"])
-		if(hasHUD(usr, EXAMINE_HUD_MEDICAL))
+		if(hasHUD(usr, EXAMINE_HUD_MEDICAL_WRITE))
 			if(usr.incapacitated())
 				return
 			var/modified = 0
```
Comment (Burzah) [inline code/modules/mob/living/carbon/human/human_mob.dm:776]:
```suggestion
			var/modified = FALSE
```
Another one to change, while we are here.

**Review by S34NW — DISMISSED** [review]

File: code/modules/mob/living/carbon/human/human_mob.dm (hunk 5 above)
Comment (Danchi299) [reply code/modules/mob/living/carbon/human/human_mob.dm:816]: judging by the medical lines of code just above, yes

Comment (Danchi299) [reply code/modules/mob/living/carbon/human/human_mob.dm:816]: after examining the code, probably not
```dm
/mob/living/carbon/human/proc/sec_hud_set_security_status()
	var/image/holder = hud_list[WANTED_HUD]
	var/perpname = get_visible_name(TRUE) //gets the name of the perp, works if they have an id or if their face is uncovered
	if(!SSticker) return //wait till the game starts or the monkeys runtime....
	if(perpname)
		var/datum/data/record/R = find_record("name", perpname, GLOB.data_core.security)
		if(R)
			switch(R.fields["criminal"])
				if(SEC_RECORD_STATUS_EXECUTE)
					holder.icon_state = "hudexecute"
					return
				if(SEC_RECORD_STATUS_ARREST)
					holder.icon_state = "hudwanted"
					return
				if(SEC_RECORD_STATUS_SEARCH)
					holder.icon_state = "hudsearch"
					return
				if(SEC_RECORD_STATUS_MONITOR)
					holder.icon_state = "hudmonitor"
					return
				if(SEC_RECORD_STATUS_DEMOTE)
					holder.icon_state = "huddemote"
					return
				if(SEC_RECORD_STATUS_INCARCERATED)
					holder.icon_state = "hudincarcerated"
					return
				if(SEC_RECORD_STATUS_PAROLLED)
					holder.icon_state = "hudparolled"
					return
				if(SEC_RECORD_STATUS_RELEASED)
					holder.icon_state = "hudreleased"
					return
	holder.icon_state = null
```

**Review by S34NW — APPROVED** [review]

**Pull request merged**